### PR TITLE
Add functionality to build yolov5 yaml

### PR DESCRIPTION
**Pull request by dsikka**
### Summary:
- Provide the functionality to build a yolov5 yaml file, given the user provides an input directory to the `data` arg.  
- The following assumptions were made about the directory structure and a sample directory structure is shown below

#### Assumptions:
- The directory passed in contains an `images` subdirectory, a `labels` subdirectory, and a file called `classes.txt` which contains the classes, ordered by class id
- Each of the `images` and `labels` subdirectories contains training, testing and validation folders
- The labels should already be in the format expected by yolov5
- yolov5 assumes the train, test and validation directories have identical names in both the `images` and `labels` folders

Sample Directory Structure:

```
- data_for_training/
    - labels/
        - train/
        - val/
        - test/
    - images/
        - train/
        - val/
        - test/
    - classes.txt
```

### Example Inputs/Outputs:

Input directory structure:

```
dsikka@quad-mle-2:~/some_dir/some_dataset$ ls
classes.txt  images
```
`classes.txt`:
```
dog
cat 
house 
park

```
`images` dir:
```
dsikka@quad-mle-2:~/some_dir/some_dataset/images$ ls
test  train  val
```

yaml file produced:

```
path: some_dataset/
train:
- images/train
test:
- images/test
val:
- images/val
names:
  0: dog
  1: cat
  2: house
  3: park
```

### Testing:
- Locally tested `training_aware` runs by passing in a local directory to the `data` arg.  yaml file produced is correct and training runs to completion

### CODE REVIEW

File: src/sparsify/auto/tasks/object_detection/yolov5/runner.py [hunk 1]
```diff
@@ -121,6 +123,105 @@ def config_to_args(
...
+                - test/
+            - classes.txt
+
+        :params dataset: inputted data string arg. Assumed to either be a dataset which
+        can be downloaded publicly or a locally available directory containing
+        data files.
+
+        :returns: path to yaml to download or the newly built yaml. If the data string
+        arg is a yaml for a publicly available dataset, this function will return the
+        same string. Otherwise, the path to the newly generated yaml will be returned.
```
Comment (rahul-tuli) [inline src/sparsify/auto/tasks/object_detection/yolov5/runner.py:156]: same missing tab

File: src/sparsify/auto/tasks/object_detection/yolov5/runner.py [hunk 2]
```diff
@@ -121,6 +123,105 @@ def config_to_args(
...
+        """
+        data_file_args = {}
+        IMAGE_DIR = "images"
+        CLASS_LIST_PATH = "classes.txt"
+        YAML_PATH = "data_local.yaml"
```
Comment (rahul-tuli) [inline src/sparsify/auto/tasks/object_detection/yolov5/runner.py:161]: nit: We only use fully UpperCased names for globals

File: src/sparsify/auto/tasks/object_detection/yolov5/runner.py [hunk 3]
```diff
@@ -121,6 +123,105 @@ def config_to_args(
...
+            with open(class_list_path, "r") as f:
+                class_list = f.readlines()
+
+            classes = {}
+            for class_id in range(len(class_list)):
+                classes[class_id] = class_list[class_id].strip()
+
+            with open(dataset, "w") as f:
+                yaml.safe_dump(
+                    {**data_file_args, **{"names": classes}}, f, sort_keys=False
```
Comment (rahul-tuli) [inline src/sparsify/auto/tasks/object_detection/yolov5/runner.py:220]: I don't think we need the inner dictionary here:

```python
>>> a = {"a": 1}
>>> b = {"b": 2, **a}
>>> b
{'b': 2, 'a': 1}
>>> c = {**a, "c": 3}
>>> c
{'a': 1, 'c': 3}
>>>
```

```suggestion
                    {**data_file_args, "names": classes}, f, sort_keys=False
```

File: src/sparsify/auto/tasks/object_detection/yolov5/runner.py [hunk 4]
```diff
@@ -121,6 +123,105 @@ def config_to_args(
...
+        CLASS_LIST_PATH = "classes.txt"
+        YAML_PATH = "data_local.yaml"
+
+        def _check_and_update_file(file_type: str, path: str):
+            if data_file_args.get(file_type, None):
+                data_file_args[file_type].append(path)
+            else:
+                data_file_args[file_type] = [path]
+
+        if os.path.isdir(dataset):
```
Comment (rahul-tuli) [inline src/sparsify/auto/tasks/object_detection/yolov5/runner.py:169]: Should we add a Log saying we're trying to generate dataset yaml from supplied dir?

File: src/sparsify/auto/tasks/object_detection/yolov5/runner.py [hunk 5]
```diff
@@ -121,6 +123,105 @@ def config_to_args(
...
+                current_path = os.path.join(IMAGE_DIR, d)
+                if re.search(r"train", d):
+                    _check_and_update_file("train", current_path)
+                elif re.search(r"val", d):
+                    _check_and_update_file("val", current_path)
+                elif re.search(r"test", d):
+                    _check_and_update_file("test", current_path)
+
+            if not (
+                data_file_args.get("train", None) and data_file_args.get("val", None)
```
Comment (rahul-tuli) [inline src/sparsify/auto/tasks/object_detection/yolov5/runner.py:199]: nit: Can remove None from get

File: src/sparsify/auto/tasks/object_detection/yolov5/runner.py (hunk 4 above)
Comment (bfineran) [reply src/sparsify/auto/tasks/object_detection/yolov5/runner.py:169]: IMO user doesn't need to care about an implementation detail like this

**Review by bfineran — COMMENTED** [review]
LGTM pending rahul's comments

File: src/sparsify/auto/tasks/object_detection/yolov5/runner.py [hunk 6]
```diff
@@ -121,6 +123,105 @@ def config_to_args(
 
         return train_args, export_args
 
+    @classmethod
+    def parse_data_args(cls, dataset: str) -> str:
```
[the remark this answers is not in the document]
Comment (dsikka) [reply src/sparsify/auto/tasks/object_detection/yolov5/runner.py:127]: moving to helpers

**Comment by dsikka** [comment]
Updated and retested.

**Review by bfineran — APPROVED** [review]

**Review by rahul-tuli — APPROVED** [review]

**Pull request merged**